rust/hyperlane-base/src/metrics/agent_metrics.rs

```diff
@@ -0,0 +1,123 @@
+    wallet_balance: Option<GaugeVec>,
+}
+
+pub(crate) fn create_agent_metrics(metrics: &CoreMetrics) -> Result<Metrics> {
+    Ok(MetricsBuilder::default()
+        .wallet_balance(metrics.new_gauge(
+            "wallet_balance",
+            WALLET_BALANCE_HELP,
+            WALLET_BALANCE_LABELS,
+        )?)
+        .build()?)
+}
+
+/// Configuration for the prometheus middleware. This can be loaded via serde.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type", rename_all = "camelCase"))]
+pub struct AgentMetricsConf {
+    /// The account to track
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub address: Option<String>,
+
+    /// Information about the chain this metric is for
+    pub domain: HyperlaneDomain,
+
+    /// Name of the agent the metrics are about
+    pub name: String,
+}
+
+/// Utility struct to update agent metrics
+#[derive(new)]
+pub struct AgentMetrics {
+    metrics: Metrics,
+    conf: AgentMetricsConf,
+    fetcher: Box<dyn AgenMetricsFetcher>,
+}
+
+impl AgentMetrics {
+    async fn update_wallet_balances(&self) {
+        let Some(wallet_addr) = self.conf.address.clone() else {
+            return;
+        };
+        let wallet_name = self.conf.name.clone();
+        let Some(wallet_balance_metric) = self.metrics.wallet_balance.clone() else {
+            return;
+        };
+        let chain = self.conf.domain.name();
+
+        match self.fetcher.get_balance().await {
+            Ok(balance) => {
+                // Okay, so the native type is not a token, but whatever, close enough.
+                // Note: This is ETH for many chains, but not all so that is why we use `N` and `Native`
+                // TODO: can we get away with scaling as 18 in all cases here? I am guessing not.
+                let balance = u256_as_scaled_f64(balance, 18);
+                trace!("Wallet {wallet_name} ({wallet_addr}) on chain {chain} balance is {balance} of the native currency");
+                wallet_balance_metric
+                    .with(&hashmap! {
+                    "chain" => chain,
+                    "wallet_address" => wallet_addr.as_str(),
+                    "wallet_name" => wallet_name.as_str(),
+                    "token_address" => "none",
+                    "token_symbol" => "Native",
+                    "token_name" => "Native"
+                }).set(balance)
+            },
+            Err(e) => warn!("Metric update failed for wallet {wallet_name} ({wallet_addr}) on chain {chain} balance for native currency; {e}")
+        }
+    }
+
+    /// Periodically updates the metrics
+    pub async fn start_updating_on_interval(self, period: Duration) {
+        let mut interval = tokio::time::interval(period);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+        loop {
+            self.update_wallet_balances().await;
+            interval.tick().await;
+        }
+    }
```

Comment:
I think there should be a way to refactor this, since this is mostly copy-pasta from the other `start_updating_on_interval` in the codebase. The only different bit is the call to `self.update_wallet_balances().await;`. However I wasn't able to come up with an async closure type that accepts both cases, since one calls from an Arc and the other one from a ref. Spent 2+ hours on this and feels bad to give up, but hopefully future Rust versions will simplify this.